### PR TITLE
Add support for declaring status code defaults

### PR DESCRIPTION
**Pull request by ritch**
/to @bajtos 
/cc @raymondfeng @dashby3000 

This adds a few new features:

 - declare a default `status` and `errorStatus` code in your `method.http` options
 - `returns.http.target`
   - `status` sets the `res.statusCode` to the provided value
   - `header` sets the `returns.header` or `returns.arg` named header to the value
 - set a custom header name for a header arg: `returns: {arg: 'length', {http: {target: 'header', header: 'Content-Length'}}}`
 
```js
describe('using a custom satus code', function() {
  it('returns a custom status code', function(done) {
    var method = givenSharedStaticMethod(
      function fn(cb) {
        cb();
      },
      {
        http: { status: 201 }
      }
    );
    json(method.url)
      .expect(201, done);
  });
  it('returns a custom error status code', function(done) {
    var method = givenSharedStaticMethod(
      function fn(cb) {
        cb(new Error('test error'));
      },
      {
        http: { status: 201, errorStatus: 508 }
      }
    );
    json(method.url)
      .expect(508, done);
  });
  it('returns a custom status code from a callback arg', function(done) {
    var exampleStatus = 222;
    var method = givenSharedStaticMethod(
      function fn(status, cb) {
        cb(null, status);
      },
      {
        accepts: { arg: 'status', type: 'number' },
        returns: {
          arg: 'status',
          http: { target: 'status' }
        }
      }
    );
    json(method.url + '?status=' + exampleStatus)
      .expect(exampleStatus, done);
  });
});
```

Connect to strongloop/strong-remoting#186

### CODE REVIEW

**Comment by ritch** [comment]
test please

File: lib/rest-adapter.js [hunk 1]
```diff
@@ -359,7 +359,9 @@ RestAdapter.errorHandler = function(options) {
         err.status = err.statusCode = 500;
       }
 
-      res.statusCode = err.statusCode || err.status || 500;
+      if (res.statusCode === undefined || res.statusCode === 200) {
```
Comment (bajtos) [inline lib/rest-adapter.js]: How about other 2xx status codes, like 204 No Content? I am proposing to rewrite the condition to `res.statusCode === undefined || res.statusCode < 400`

File: lib/http-context.js [hunk 2]
```diff
@@ -332,7 +336,41 @@ HttpContext.prototype.invoke = function(scope, method, fn, isCtor) {
...
+  if (returnDesc.root) {
+    this.result = value;
+  } else if (returnDesc.http) {
+    switch (returnDesc.http.target) {
+      case 'status':
+        res.status(value);
+        break;
+      case 'header':
+        res.set(returnDesc.header || name, value);
+        break;
```
Comment (bajtos) [inline lib/http-context.js]: Could you please add a unit-test for this?

**Comment by bajtos** [comment]
I am not sure if it's a good idea to add a constant `errorStatus` code, I am concerned that it may mask useful status codes like 422 (validation failed), 400 (bad request, caused by a missing argument) etc. Could you please add a unit- test to verify that `err.statusCode` takes precedence over `errorStatus` defined in remoting metadata?

```js
       var method = givenSharedStaticMethod(
         function fn(cb) {
           var err = new Error('test error');
           err.statusCode = 400;
           cb(err);
         },
         {
           http: { status: 201, errorStatus: 508 }
         }
       );
       json(method.url)
         .expect(400, done);
```

**Comment by bajtos** [comment]
Awesome stuff!

It seems to me that passing `ctx` to `SharedMethod.prototype.invoke` is a bit inelegant and creates coupling that may make future refactoring more difficult. I don't have a better solution at hand though, so don't worry about it too much.

**Comment by ritch** [comment]
> I am not sure if it's a good idea to add a constant errorStatus code, I am concerned that it may mask useful status codes like 422 (validation failed), 400 (bad request, caused by a missing argument) etc. 

Well we need some way of declaring the status code for a specific result (eg. when the result is ok send this status, when the result is an error send this status).

I think the `err.status` should take precedence. So if that is set we will just ignore the `errorStatus` default.

**Comment by crandmck** [comment]
@ritch Remote methods' accepts & returns are documented here: http://apidocs.strongloop.com/strong-remoting/#sharedmethod (in source file https://github.com/strongloop/strong-remoting/blob/master/lib/shared-method.js). It might be best for you to add docs, since I'm not familiar with this PR.  But if you want me to do it, please just open a doc issue.

File: lib/http-context.js [hunk 3]
```diff
@@ -384,6 +424,11 @@ HttpContext.prototype.done = function() {
   var data = this.result;
   var res = this.res;
   var accepts = this.req.accepts(this.supportedTypes);
+  var defaultStatus = this.method.http.status;
+
+  if (defaultStatus) {
+    res.status(defaultStatus);
```
Comment (bajtos) [inline lib/http-context.js]: Should we check that `res.statusCode === undefined || res.statusCode === 200` before applying `defaultStatus`?

**Comment by bajtos** [comment]
The code LGTM, although I left one comment/question for your consideration. Please make sure the CI builds are green before landing this patch.

File: lib/http-context.js (hunk 3 above)
Comment (ritch) [inline lib/http-context.js]: Added the check...

**Comment by auguster** [comment]
I'm having a use case with long video processing, sometimes the video is ready (status code 200) or processing (202). You can get the detailed status on a specific method that returns a status 201 if the processing has finished with the URI on where to get it.

So far I'm setting the status code through `new Error(...).statusCode` the problem is that 2xx are not errors and strong-error-handler is replacing status code < 400 by 500.

What has been proposed so far (setting static return status) doesn't work in my use case as I can return 200 or 202.

I'm looking for the best way to return non-error status codes. I guess I could go into the ctx.res and manually set this but that feels so wrong. What should I do ?

**Comment by bajtos** [comment]
@auguster Your method should have a callback argument that's annotated with `http: {target: 'status'}`. See https://loopback.io/doc/en/lb3/Remote-methods.html#argument-descriptions
 and the test-cases checking this functionality:

https://github.com/strongloop/strong-remoting/blob/4ef9baac55a5c35b5b6f8f2252bf0699f758e837/test/rest.test.js#L2052-L2095

**Comment by auguster** [comment]
@bajtos I will try that, thanks.

**Pull request merged**